rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,337 @@
+framework will automatically make those jobs `pending` after a timeout. That
+ensures the jobs continue to make progress. To avoid getting re-enqueued as
+`pending` due the timeout, each job processor must periodically call the
+`update/2,3` function. That functionality is implemented by the `activity
+monitor`. It periodically watches a per-type versionstamp-ed key, then scans
+`active jobs` subspace for any `running` jobs which haven't updated their
+entries during the timeout period.
+
+### Subscription Notifications
+
+Subscription notifications are managed separately for each job type. They use
+a per-type versionstamp-ed watch to monitor which jobs have updated since
+the last time it delivered notifications to the subscribers.
+
+### Data Model
+
+ * `("couch_jobs", "data", Type, JobId) = (Sequence, JobLock, ScheduledTime, Resubmit, JobData)`
+ * `("couch_jobs", "pending", Type, ScheduledTime, JobId) = ""`
+ * `("couch_jobs", "watches", "pending", Type) = (Sequence, Counter)`
+ * `("couch_jobs", "watches", "activity", Type) = Sequence`
+ * `("couch_jobs", "activity_timeout", Type) = ActivityTimeout`
+ * `("couch_jobs", "activity", Type, Sequence) = JobId`
+
+
+### Job Lifecycle Implementation
+
+This section describes how the framework implements some of the API functions.
+
+ - `add/3,4` :
+   * Add the new job to the main jobs table
+   * If a job with the same `JobId` exists, return an error
+   * Update `"pending"` watch for the type with a new versionstamp and bump its
+     counter.
+   * `JobLock` is set to `null`.
+
+ - `remove/1` :
+   * Job is removed from the main jobs table
+   * Job processor during the next `update/2,3` call will get a `halt` error
+     and know to stop running the job.
+
+ - `accept/1,2` :
+   * Generate a unique `JobLock` UUID.
+   * Attempt to dequeue the item from the pending queue, then assign it the
+     `JobLock` in the jobs table.
+   * Create entry in the `"activity"` subspace.
+   * If there are no pending jobs, get a watch for the `"pending"` queue and
+     wait until it fires, then try again.
+
+ - `update/2,3`:
+   * If job is missing from the main jobs table return `halt`
+   * Check if `JobLock` matches, otherwise return `halt`
+   * Delete old `"activity"` sequence entry
+   * Maybe update `JobData`
+   * Create a new `"activity"` sequence entry and in main job table
+   * Update `"watches"` sequence for that job type
+
+ - `finish/2,3`:
+   * If job is missing from the main jobs table return `halt`
+   * Check if `JobLock` matches, otherwise returns `halt`
+   * Delete old `"activity"` sequence entry
+   * If `Resubmit` field is `true`, re-enqueue the job, and set `Resubmit` to `false`.
+   * Set job table's `JobLock` to `null`
+
+ - `resubmit/1`:
+   * Set the `Resubmit` field to `true`
+   * The job will be re-enqueued by the `finish/2,3` call
+
+# Advantages and Disadvantages
+
+A previous draft of this RFC discussed an implementation centered around
```

Comment:
I don't think this paragraph is really relevant. The advantages and disadvantages should be focused around this design.  I think the advantages are more along the lines that we have a central way to coordinate batch processing across a cluster. A nice unified api. Not sure around disadvantages

Reply:
Agree.